examples/hardware_encoding/main.go

```diff
@@ -0,0 +1,204 @@
+		return
+	}
+
+	// Open output file
+	output, err := os.OpenFile(*output, os.O_CREATE|os.O_WRONLY, 0o644)
+	if err != nil {
+		log.Fatal(fmt.Errorf("main: opening output file failed: %w", err))
+	}
+	defer output.Close()
+
+	// Get hardware device type
+	hardwareDeviceType := astiav.FindHardwareDeviceTypeByName(*hardwareDeviceTypeName)
+	if hardwareDeviceType == astiav.HardwareDeviceTypeNone {
+		log.Fatal(errors.New("main: hardware device not found"))
+	}
+
+	// Create hardware device context
+	hardwareDeviceContext, err := astiav.CreateHardwareDeviceContext(hardwareDeviceType, *hardwareDeviceName, nil)
+	if err != nil {
+		log.Fatal(fmt.Errorf("main: creating hardware device context failed: %w", err))
+	}
+
+	// Find encoder codec
+	encCodec := astiav.FindEncoderByName(*encoderCodecName)
+	if encCodec == nil {
+		log.Fatal("main: encoder codec is nil")
+	}
+
+	// Alloc codec context
+	encCodecContext := astiav.AllocCodecContext(encCodec)
+	if encCodecContext == nil {
+		log.Fatal("main: codec context is nil")
+	}
+	defer encCodecContext.Free()
+
+	// Set codec context
+	encCodecContext.SetWidth(*width)
+	encCodecContext.SetHeight(*height)
+	encCodecContext.SetTimeBase(astiav.NewRational(1, *fps))
+	encCodecContext.SetFramerate(astiav.NewRational(*fps, 1))
+	hardwarePixelFormatName := astiav.FindPixelFormatByName(*hardwarePixelFormatName)
+	if hardwarePixelFormatName == astiav.PixelFormatNone {
+		log.Fatal("main: hardware pixel format not found")
+	}
+	encCodecContext.SetPixelFormat(hardwarePixelFormatName)
+
+	// Set hardware frame context
+	hardwareFrameCtx := astiav.AllocHardwareFrameContext(hardwareDeviceContext)
+	if hardwareFrameCtx == nil {
+		log.Fatal("main: hardware frame context is nil")
+	}
+	hardwareFrameCtx.SetPixelFormat(hardwarePixelFormatName)
+	hardwareFrameCtx.SetSoftwarePixelFormat(astiav.PixelFormatNv12)
+	hardwareFrameCtx.SetWidth(*width)
+	hardwareFrameCtx.SetHeight(*height)
+	hardwareFrameCtx.SetInitialPoolSize(*initialPoolSize)
+
+	// Initialize hardware frame context
+	if err := hardwareFrameCtx.Initialize(); err != nil {
+		log.Fatal(fmt.Errorf("main: initializing hardware frame context failed: %w", err))
+	}
+	encCodecContext.SetHardwareFrameContext(hardwareFrameCtx)
+
+	// Open codec context
+	if err := encCodecContext.Open(encCodec, nil); err != nil {
+		log.Fatal(fmt.Errorf("main: opening codec context failed: %w", err))
+	}
+
+	frameIndex := 0
+
+	// Draw frames, upload them to hardware devices, and encode them
+	for {
+		// Alloc software frame
+		softwareFrame := astiav.AllocFrame()
+
+		// Set software frame
+		softwareFrame.SetWidth(*width)
+		softwareFrame.SetHeight(*height)
+		softwareFrame.SetPixelFormat(astiav.PixelFormatNv12)
+
+		// Alloc software frame buffer
+		if err := softwareFrame.AllocBuffer(0); err != nil {
+			log.Fatal(fmt.Errorf("main: allocating buffer failed: %w", err))
+		}
+
+		// Fill software frame
```

Comment:
Although your idea is really cool, I'd rather keep things simple regarding the source frame and only do
```go
// Fill software frame with black
if err = srcFrame.ImageFillBlack(); err != nil {
	log.Fatal(fmt.Errorf("main: filling source frame with black image failed: %w", err))
}
```